### PR TITLE
feat(git): automatically skip `git status` for huge repo

### PR DESCRIPTION
**Pull request by bixycler**
## Standards checklist:

- [x] The PR title is descriptive.
- [x] The PR doesn't replicate another PR which is already open.
- [x] I have read the contribution guide and followed all the instructions.
- [x] The code follows the code style guide detailed in the wiki.
- [x] The code is mine or it's from somewhere with an MIT-compatible license.
- [x] The code is efficient, to the best of my ability, and does not waste computer resources.
- [x] The code is stable and I have tested it myself, to the best of my abilities.

## Overview:
There have been already two flags `oh-my-zsh.hide-status` and `oh-my-zsh.hide-dirty` for user to *manually* set for each heavy repo to prevent calling `git status` in those repos to improve performance. However, we can easily ***automate*** this performance improvement by checking the number of indexed files against a threshold. Moreover, we can dynamically change that threshold in the terminal whenever we see a notable change in performance.

- This PR subsumes the PR #10897

## New features:
- `git_repo_files_number()` prints the number of indexed files in this repo;
  returns 0 or error code: 128 = "not in git repo", 2 = "index read error",
  3 = "sharedindex read error"
- `ZSH_THEME_GIT_HUGE_REPO_THRESHOLD`: when there are more indexed files
  than this number (default 100 000), the repo is considered "huge"
- `git_huge_repo()` returns 0 when this repo is "huge", 1 when this repo
  is not "huge", otherwise the error number returned from `git_repo_files_number()`
- `ZSH_THEME_GIT_PROMPT_HUGE` is the prompt text indicating a "huge" repo, 
  e.g. `ZSH_THEME_GIT_PROMPT_HUGE="Ω"`
  This can be used together with `git_repo_files_number()`, 
  e.g. `"$(git_prompt_status)($(git_repo_files_number) files)"` => `"Ω(396604 files)"`

## Changes:

- `git_prompt_status()` prints `ZSH_THEME_GIT_PROMPT_HUGE` and does not
  call `git status` (returns 2) for "huge" repo
- `parse_git_dirty()` returns immediately (prints nothing) for "huge" repo

### System info:
| | |
| ----------- | ----------- |
| Model: | iMac | 
| CPU: | Apple M1, 8 cores |
| RAM: | 8 GB |
| OS: | MacOS Monterey |

### Performance (`time`) for a 396,604-file repo (`chromium`):

| as a sub-huge repo | ZSH_THEME_GIT_HUGE_REPO_THRESHOLD=1000000 |
| ----------- | ----------- |
| git status              | 0.36s user 4.72s system 64% cpu 7.890 total |
| ( git_prompt_status; )  | 0.37s user 4.66s system 64% cpu 7.828 total |
| ( parse_git_dirty; )    | 0.37s user 4.70s system 63% cpu 7.913 total |
| ( git_prompt_info; )    | 0.37s user 4.71s system 65% cpu 7.799 total |
| ( git_repo_files_number; )  | 0.01s user 0.01s system 73% cpu 0.032 total |
| ( git_huge_repo; )          | 0.01s user 0.01s system 75% cpu 0.034 total |

| as a huge repo | ZSH_THEME_GIT_HUGE_REPO_THRESHOLD=100000 |
| ----------- | ----------- |
| git status              | 0.36s user 4.77s system 63% cpu 8.094 total |
| ( git_prompt_status; )  | 0.01s user 0.02s system 74% cpu 0.042 total |
| ( parse_git_dirty; )    | 0.01s user 0.01s system 75% cpu 0.030 total |
| ( git_prompt_info; )    | 0.02s user 0.03s system 74% cpu 0.064 total |
| ( git_repo_files_number; )  | 0.01s user 0.01s system 74% cpu 0.032 total |
| ( git_huge_repo; )          | 0.01s user 0.01s system 71% cpu 0.032 total |

### CODE REVIEW

**Comment by bixycler** [comment]
@mcornella please give it a look!

**Comment by psztoch** [comment]
Problem with `git status` performance is especially noticeable on the Windows Linux Subsystem.

File: lib/git.zsh [hunk 1]
```diff
@@ -164,8 +168,46 @@ function git_prompt_long_sha() {
...
+  local X
+  local N
+  #N=$(__git_prompt_git ls-files 2> /dev/null | wc -l) # `git ls-files` is much faster than `git status` but still too slow!
+  gitdir=$(__git_prompt_git rev-parse --git-dir 2> /dev/null)
+  [[ $? -eq 128 ]] && return 128
+  [[ "$(__git_prompt_git config --get oh-my-zsh.hide-info 2>/dev/null)" == 1 ]] && return 0
+  # parse index file(s) [https://git-scm.com/docs/index-format]
+  X=$(xxd -p -l4 -s12 "$gitdir/$indexf" 2>/dev/null) # extension signature
+  [[ $? -ne 0 ]] && return 2
+  if [[ "$X" == "6c696e6b" ]]; then # "link" extension --> split/shared index
```
Comment (jmeridth) [inline lib/git.zsh:183]: Might be because I'm on my phone but why the specific SHA? Thanks for clarifying.

Comment (bixycler) [reply lib/git.zsh:183]:
```
# parse index file(s) [https://git-scm.com/docs/index-format]
  X=$(xxd -p -l4 -s12 "$gitdir/$indexf" 2>/dev/null) # extension signature
  [[ $? -ne 0 ]] && return 2
  if [[ "$X" == "6c696e6b" ]]; then # "link" extension --> split/shared index
```
As shown in [index-format](https://git-scm.com/docs/index-format):

- After the 12-byte header is a 4-byte extension signature, hence `X=$(xxd -p -l4 -s12 ...)`
- The [split index](https://git-scm.com/docs/index-format#_split_index) has an extension signature literally spelled "`link`" in ASCII whose HEX value is `0x6c696e6b`, hence `[[ "$X" == "6c696e6b" ]]`. We can use `xxd -r` to convert it back from HEX to ASCII, but it's obviously superfluous.

**Comment by karczex** [comment]
@bixycler  Your solution with parsing git internal data seems to be little bit to fragile to changes in git internals. Also definition of that what is a huge repo is user's hardware dependent. Based on hints from git status and https://github.com/ohmyzsh/ohmyzsh/pull/9630 I tried such approach: https://github.com/karczex/ohmyzsh/commits/parse_git_dirty_speedup and it works pretty well for me. Are you able to compare it with your solution (on your hardware)?

**Comment by larson-carter** [comment]
Echoing what @karczex said. I do believe that this could be dependent on git internals. 

Aside from that - I do think that if we had more benchmarks, as well as a standard for the threshold of what becomes "slow" that could help us further develop this.

**Comment by carlosala** [comment]
Closing this PR. We have a new system now that makes the `git status` to be async 🎉 
It shouldn't be needed anymore!

**Comment by larson-carter** [comment]
Good point, I will need to test this on a few personal repos to see how it performs.

I would like to do some research/data analysis to see how it performs.

**Comment by YodaEmbedding** [comment]
For non-async users, `git_prompt_info` can become very slow -- 3-8 seconds.

Running `git status` every once in a while refreshes the index, and this seems to help speed things up. (`git_prompt_info` still takes a second, even then.)

**Pull request closed without merging**